system/Model.php

```diff
@@ -723,7 +730,21 @@ public function insert($data = null, bool $returnID = true)
 		// If insertion succeeded then save the insert ID
 		if ($result)
 		{
-			$this->insertID = $this->db->insertID();
+			if (! $this->useAutoIncrement)
+			{
+				if (empty($data[$this->primaryKey]))
+				{
+					$this->insertID = null;
+				}
+				else
+				{
+					$this->insertID = $data[$this->primaryKey];
+				}
+			}
```

Comment:
Instead of using `$data` array we should use `$eventData`. This way developers can take advantage of model filers.
```suggestion
			if (! $this->useAutoIncrement)
			{
				$this->insertID = $eventData['data'][$this->primaryKey];
			}
```
Also, we need to check if there is a primary key set when `$useAutoIncrement === false`, so right after `beforeInsert` (line 723), we should add a check, like:

```php
$eventData = $this->trigger('beforeInsert', ['data' => $data]);

if (! $this->useAutoIncrement && empty($eventData['data'][$this->primaryKey]))
{
        throw DataException::forEmptyPrimaryKey();
}
```
We need to create a new method in `system/Database/Exceptions/DataException.php` file and also an entry to the language file - something like "You must provide a Primary Key value". If you're not sure how to handle this, just see how other exceptions in this class look like.